node/overseer/src/lib.rs

```diff
@@ -725,68 +763,159 @@ where
 		let metrics = self.metrics.clone();
 		spawn_metronome_metrics(&mut self, metrics)?;
 
-		// Notify about active leaves on startup before starting the loop
+		let initial_sync_finished = self.sync_oracle.finished_syncing();
+		// Import the active leaves found in the database
 		for (hash, number) in std::mem::take(&mut self.leaves) {
 			let _ = self.active_leaves.insert(hash, number);
 			if let Some((span, status)) = self.on_head_activated(&hash, None).await {
-				let update =
-					ActiveLeavesUpdate::start_work(ActivatedLeaf { hash, number, status, span });
-				self.broadcast_signal(OverseerSignal::ActiveLeaves(update)).await?;
+				if initial_sync_finished {
+					// Initial sync is complete. Notify the subsystems and proceed to the main loop
+					let update = ActiveLeavesUpdate::start_work(ActivatedLeaf {
+						hash,
+						number,
+						status,
+						span,
+					});
+					self.broadcast_signal(OverseerSignal::ActiveLeaves(update)).await?;
+				}
 			}
 		}
 
-		loop {
+		// If initial sync is not complete - wait for it.
+		// This loop is identical to the main one but doesn't generate `ActiveLeaves` and `BlockFinalized` events until
+		// the initial full sync is complete.
+		// This is an infinite loop which executes only when `!initial_sync_finished`. The weird syntax is only to save
+		// one extra layer if indentation because it's already a bit tough for the eyes.
+		// Think about it like:
+		// ```
+		// if !initial_sync_finished {
+		//		loop {
+		//			select! {
+		//				...
+		//			}
+		// 		}
+		// }
+		//```
+		while !initial_sync_finished {
 			select! {
 				msg = self.events_rx.select_next_some() => {
 					match msg {
-						Event::MsgToSubsystem { msg, origin } => {
-							self.route_message(msg.into(), origin).await?;
-							self.metrics.on_message_relayed();
-						}
-						Event::Stop => {
-							self.stop().await;
-							return Ok(());
-						}
-						Event::BlockImported(block) => {
-							self.block_imported(block).await?;
-						}
-						Event::BlockFinalized(block) => {
-							self.block_finalized(block).await?;
-						}
-						Event::ExternalRequest(request) => {
-							self.handle_external_request(request);
-						}
+						Event::MsgToSubsystem { msg, origin } => self.handle_msg_to_subsystem(msg, origin).await?,
+						Event::Stop => return self.handle_stop().await,
+						Event::BlockImported(block) => _ = self.block_imported(block).await,
+						Event::BlockFinalized(block) if self.sync_oracle.finished_syncing() => {
+							// Initial sync is complete
```

Comment:
Wouldn't we send active_leaves twice if we receive finished_syncing -> BlockImported -> BlockFinalized?

Reply:
I think we won't. The only way to break out of this loop is via `BlockFinalized`. `self.block_imported(block).await` won't send any event. Or I am missing something?